File: plugins/ads/source/iOS/Plugins/Ads/frameworks/AdsChartboost.m
//
//  AdsChartboost.m
//  Ads
//
//  Created by Arturs Sosins on 6/25/13.
//  Copyright (c) 2013 Gideros Mobile. All rights reserved.
//
#include "gideros.h"
#import "AdsChartboost.h"
#import "AdsClass.h"
#import <Chartboost/CBNewsfeedUI.h>

@implementation AdsChartboost
-(id)init{
    self.mngr = [[AdsManager alloc] init];
    return self;
}

-(void)destroy{
    [self.mngr destroy];
    [self.mngr release];
    self.mngr = nil;
}

-(void)setKey:(NSMutableArray*)parameters{
    [Chartboost startWithAppId:[parameters objectAtIndex:0] appSignature:[parameters objectAtIndex:1] delegate:self];
    [CBNewsfeed startWithDelegate:self];
}

-(void)loadAd:(NSMutableArray*)parameters{
    NSString *type = [parameters objectAtIndex:0];
    NSString *tag = CBLocationDefault;
    if([parameters count] > 1)
        tag = [parameters objectAtIndex:1];
    if ([type isEqualToString:@"interstitial"] || [type isEqualToString:@"auto"]) {
        AdsStateChangeListener *listener = [[AdsStateChangeListener alloc] init];
        [listener setShow:^(){
            [AdsClass adDisplayed:[self class] forType:type];
            [Chartboost showInterstitial:tag];
        }];
        [listener setDestroy:^(){}];
        [listener setHide:^(){}];
        [self.mngr set:listener forType:type withListener:listener];
        if([Chartboost hasInterstitial:tag]){
            [self.mngr load:type];
            [AdsClass adReceived:[self class] forType:type];
        }
        else
            [Chartboost cacheInterstitial:tag];
    }
    else if ([type isEqualToString:@"moreapps"]) {
        AdsStateChangeListener *listener = [[AdsStateChangeListener alloc] init];
        [listener setShow:^(){
            [AdsClass adDisplayed:[self class] forType:type];
            [Chartboost showMoreApps:tag];
        }];
        [listener setDestroy:^(){}];
        [listener setHide:^(){}];
        [self.mngr set:listener forType:type withListener:listener];
        if([Chartboost hasMoreApps:tag]){
            [self.mngr load:type];
            [AdsClass adReceived:[self class] forType:type];
        }
        else
            [Chartboost cacheMoreApps:tag];
    }
    else if ([type isEqualToString:@"feed"]) {
        AdsStateChangeListener *listener = [[AdsStateChangeListener alloc] init];
        [listener setShow:^(){
            [AdsClass adDisplayed:[self class] forType:type];
            if([tag isEqualToString:@"notification"])
                [CBNewsfeed showNotificationUI];
            else
                [CBNewsfeed showNewsfeedUI];
        }];
        [listener setDestroy:^(){}];
        [listener setHide:^(){
            if([tag isEqualToString:@"notification"])
                [CBNewsfeed closeNotificationUI];
            else
                [CBNewsfeed closeNewsfeedUI];
        }];
        [self.mngr set:listener forType:type withListener:listener];
        [self.mngr load:type];
        [AdsClass adReceived:[self class] forType:type];
    }
    else if ([type isEqualToString:@"v4vc"]) {
        AdsStateChangeListener *listener = [[AdsStateChangeListener alloc] init];
        [listener setShow:^(){
            [AdsClass adDisplayed:[self class] forType:type];
            [Chartboost showRewardedVideo:tag];
        }];
        [listener setDestroy:^(){}];
        [listener setHide:^(){}];
        [self.mngr set:listener forType:type withListener:listener];
        if([Chartboost hasRewardedVideo:tag]){
            [self.mngr load:type];
            [AdsClass adReceived:[self class] forType:type];
        }
        else
            [Chartboost cacheRewardedVideo:tag];
    }
    else
    {
         [AdsClass adError:[self class] with:[NSString stringWithFormat:@"Unknown type: %@", type]];
    }
}

-(void)showAd:(NSMutableArray*)parameters{
    NSString *type = [parameters objectAtIndex:0];
    if([self.mngr get:type] == nil)
        [self loadAd:parameters];
    [self.mngr show:type];
}

-(void)hideAd:(NSString*)type{
    [self.mngr hide:type];
}

-(void)enableTesting{
}

-(UIView*)getView{
    return nil;
}

- (BOOL)shouldRequestInterstitial:(NSString *)location{
    return YES;
}

- (BOOL)shouldDisplayInterstitial:(NSString *)location{
    return YES;
}

- (void)didCacheInterstitial:(NSString *)location{
    if([self.mngr get:@"interstitial"] != nil && ![self.mngr isLoaded:@"interstitial"]){
        [AdsClass adReceived:[self class] forType:@"interstitial"];
        [self.mngr load:@"interstitial"];
    }
}

- (void)didFailToLoadInterstitial:(NSString *)location{
    [AdsClass adFailed:[self class] with:@"Failed to receive ads" forType:@"interstitial"];
    [self.mngr reset:@"interstitial"];
}

- (void)didDismissInterstitial:(NSString *)location{
    [AdsClass adDismissed:[self class] forType:@"interstitial"];
}

- (void)didCloseInterstitial:(NSString *)location{
    [AdsClass adActionEnd:[self class] forType:@"interstitial"];
}

- (void)didClickInterstitial:(NSString *)location{
    [AdsClass adActionBegin:[self class] forType:@"interstitial"];
}


- (void)didCacheMoreApps:(NSString *)location{
    if([self.mngr get:@"moreapps"] != nil && ![self.mngr isLoaded:@"moreapps"]){
        [AdsClass adReceived:[self class] forType:@"moreapps"];
        [self.mngr load:@"moreapps"];
    }

}

- (BOOL)shouldDisplayMoreApps:(NSString *)location{
    return YES;
}

- (void)didFailToLoadMoreApps:(CBLocation)location withError:(CBLoadError)error{
    [AdsClass adFailed:[self class] with:@"Failed to receive ads" forType:@"moreapps"];
    [self.mngr reset:@"moreapps"];
}

- (void)didDismissMoreApps:(NSString *)location{
    [AdsClass adDismissed:[self class] forType:@"moreapps"];
}

- (void)didCloseMoreApps:(NSString *)location{
    [AdsClass adActionEnd:[self class] forType:@"moreapps"];
}

- (void)didClickMoreApps:(NSString *)location{
    [AdsClass adActionBegin:[self class] forType:@"moreapps"];
}

// Called before a rewarded video will be displayed on the screen.
- (BOOL)shouldDisplayRewardedVideo:(CBLocation)location{
    return YES;
}

// Called after a rewarded video has been loaded from the Chartboost API
// servers and cached locally.
- (void)didCacheRewardedVideo:(CBLocation)location{
    if([self.mngr get:@"v4vc"] != nil && ![self.mngr isLoaded:@"v4vc"]){
        [AdsClass adReceived:[self class] forType:@"v4vc"];
        [self.mngr load:@"v4vc"];
    }
}

// Called after a rewarded video has attempted to load from the Chartboost API
// servers but failed.
- (void)didFailToLoadRewardedVideo:(CBLocation)location withError:(CBLoadError)error{
    [AdsClass adFailed:[self class] with:@"Failed to receive ads" forType:@"v4vc"];
    [self.mngr reset:@"v4vc"];
}

// Called after a rewarded video has been dismissed.
- (void)didDismissRewardedVideo:(CBLocation)location{
    [AdsClass adDismissed:[self class] forType:@"v4vc"];
}

// Called after a rewarded video has been clicked.
- (void)didClickRewardedVideo:(CBLocation)location{
    [AdsClass adActionBegin:[self class] forType:@"v4vc"];
}

// Called after a rewarded video has been viewed completely and user is eligible for reward.
- (void)didCompleteRewardedVideo:(CBLocation)location withReward:(int)reward{
    [AdsClass adActionEnd:[self class] forType:@"v4vc"];
}

@end

File: plugins/ads/source/Android/src/com/giderosmobile/android/plugins/ads/frameworks/AdsChartboost.java
package com.giderosmobile.android.plugins.ads.frameworks;

import java.lang.ref.WeakReference;
import android.app.Activity;
import android.util.SparseArray;
import android.view.KeyEvent;

import com.chartboost.sdk.CBLocation;
import com.chartboost.sdk.Chartboost;
import com.chartboost.sdk.ChartboostDelegate;
import com.chartboost.sdk.Model.CBError.CBImpressionError;
import com.giderosmobile.android.plugins.ads.*;

public class AdsChartboost extends ChartboostDelegate implements AdsInterface{
	
	private WeakReference<Activity> sActivity;
	private String adsID;
	private AdsManager mngr;
	static AdsChartboost me;
	
	public void onCreate(WeakReference<Activity> activity)
	{
		me = this;
		sActivity = activity;
		mngr = new AdsManager();
	}
	
	//on destroy event
	public void onDestroy()
	{	
		Chartboost.onDestroy(sActivity.get());
		mngr.destroy();
	}
	
	public void onStart(){}

	public void onStop()
	{
		Chartboost.onStop(sActivity.get());
	}
	
	public void onPause(){
		Chartboost.onPause(sActivity.get());
	}
		
	public void onResume(){
		Chartboost.onResume(sActivity.get());
	}
	
	public boolean onKeyUp(int keyCode, KeyEvent event) {
		if (Chartboost.onBackPressed())
			// If a Chartboost view exists, close it and return
			return true;
		else
			// If no Chartboost view exists, continue on as normal
			return false;
	}
	
	public void setKey(final Object parameters){
		SparseArray<String> param = (SparseArray<String>)parameters;
		adsID = param.get(0);
		Chartboost.startWithAppId(sActivity.get(), adsID, param.get(1));
		Chartboost.setImpressionsUseActivities(true);
		Chartboost.setDelegate(this);
		Chartboost.onCreate(sActivity.get());
		Chartboost.onStart(sActivity.get());
		Chartboost.setShouldDisplayLoadingViewForMoreApps(true);
	}
	
	//load an Ad
	public void loadAd(final Object parameters)
	{
		SparseArray<String> param = (SparseArray<String>)parameters;
		final String type = param.get(0);
		String second = param.get(1);
		if(second == null)
			second = CBLocation.LOCATION_DEFAULT;
		final String tag = second;
		if(type.equals("interstitial")){
			mngr.set(Chartboost.class, type, new AdsStateChangeListener(){
				@Override
				public void onShow() {
					Ads.adDisplayed(me, type);
					Chartboost.showInterstitial(tag);
				}

				@Override
				public void onDestroy() {}	
				@Override
				public void onHide() {}	
			});
			if(Chartboost.hasInterstitial(tag)){
				mngr.load(type);
				Ads.adReceived(me, type);
			}
			else
				Chartboost.cacheInterstitial(tag);
		}
		else if(type.equals("moreapps")){
			mngr.set(Chartboost.class, type, new AdsStateChangeListener(){
				@Override
				public void onShow() {
					Ads.adDisplayed(me, type);
					Chartboost.showMoreApps(tag);
				}

				@Override
				public void onDestroy() {}	
				@Override
				public void onHide() {}	
			});
			if(Chartboost.hasMoreApps(tag)){
				mngr.load(type);
				Ads.adReceived(me, type);
			}
			else
				Chartboost.cacheMoreApps(tag);
		}
		else if(type.equals("v4vc")){
			mngr.set(Chartboost.class, type, new AdsStateChangeListener(){
				@Override
				public void onShow() {
					Ads.adDisplayed(me, type);
					Chartboost.showRewardedVideo(tag);
				}

				@Override
				public void onDestroy() {}	
				@Override
				public void onHide() {}	
			});
			if(Chartboost.hasRewardedVideo(tag)){
				mngr.load(type);
				Ads.adReceived(me, type);
			}
			else
				Chartboost.cacheRewardedVideo(tag);
		}
		else
		{
			Ads.adError(this, "Unknown type: " + type);
		}
	}
	
	public void showAd(final Object parameters)
	{
		SparseArray<String> param = (SparseArray<String>)parameters;
		String type = param.get(0);
		if(mngr.get(type) == null)
		{
			loadAd(parameters);
		}
		mngr.show(type);
	}
	
	//remove ad
	public void hideAd(String type)
	{
		mngr.hide(type);
	}
	
	public int getWidth(){
		return 0;
	}
	
	public int getHeight(){
		return 0;
	}

	@Override
	public void didDismissInterstitial(String arg0) {
		Ads.adDismissed(this, "interstitial");
	}

	@Override
	public void didDismissMoreApps(String tag) {
		Ads.adDismissed(this, "moreapps");
	}
	
	 @Override
     public void didDismissRewardedVideo(String location) {
		 Ads.adDismissed(this, "v4vc");
     }
	
	@Override
	public void didCloseMoreApps(String tag) {
		Ads.adActionEnd(this, "moreapps");
	}

	
	@Override
	public void didCloseInterstitial(String arg0) {
		Ads.adActionEnd(this, "interstitial");
	}
	
	@Override
    public void didCompleteRewardedVideo(String location, int reward) {
       Ads.adActionEnd(this, "v4vc");
    }
	
	@Override
	public void didCacheInterstitial(String arg0) {
		if(mngr.get("interstitial") != null && !mngr.isLoaded("interstitial")){
			mngr.load("interstitial");
			Ads.adReceived(this, "interstitial");
		}
	}

	@Override
	public void didCacheMoreApps(String tag) {
		if(mngr.get("moreapps") != null && !mngr.isLoaded("moreapps")){
			mngr.load("moreapps");
			Ads.adReceived(this, "moreapps");
		}
	}
	
	@Override
    public void didCacheRewardedVideo(String location) {
		if(mngr.get("v4vc") != null && !mngr.isLoaded("v4vc")){
			mngr.load("v4vc");
			Ads.adReceived(this, "v4vc");
		}
    }

	@Override
	public void didClickInterstitial(String arg0) {
		Ads.adActionBegin(this, "interstitial");
	}

	@Override
	public void didClickMoreApps(String tag) {
		Ads.adActionBegin(this, "moreapps");
	}
	
	@Override
	public boolean shouldRequestInterstitial(String arg0) {
		return true;
	}

	@Override
	public boolean shouldDisplayInterstitial(String arg0) {
		return true;
	}

	@Override
	public boolean shouldDisplayMoreApps(String tag) {
		return true;
	}
	
	@Override
	public boolean shouldRequestMoreApps(String tag) {
		return true;
	}
	
	@Override
    public boolean shouldDisplayRewardedVideo(String location) {
        return true;
    }

	@Override
	public void enableTesting() {
		//should be enabled inside chartboost account
		//String android_id = Secure.getString(sActivity.get().getBaseContext().getContentResolver(),Secure.ANDROID_ID);
		//Log.e("Chartboost", android_id);
	}

	@Override
	public void didFailToLoadInterstitial(String arg0, CBImpressionError error) {
		Ads.adFailed(this, "interstitial", error.name());
		mngr.reset("interstitial");
	}

	@Override
	public void didFailToLoadMoreApps(String tag, CBImpressionError error) {
		Ads.adFailed(this, "moreapps", error.name());
		mngr.reset("moreapps");
	}
	
	@Override
    public void didFailToLoadRewardedVideo(String location, CBImpressionError error) {
		Ads.adFailed(this, "v4vc", error.name());
		mngr.reset("v4vc");
    }
}
